File: src/main/java/com/now/nowbot/service/MessageServiceImpl/TestTaikoSRCalculateService.java
package com.now.nowbot.service.MessageServiceImpl;

import com.now.nowbot.qq.event.MessageEvent;
import com.now.nowbot.service.MessageService;
import com.now.nowbot.util.Instructions;
import org.springframework.stereotype.Service;

import java.util.LinkedList;
import java.util.List;
import java.util.regex.Matcher;

@Service("TEST_TAIKO_SR_CALCULATE")
public class TestTaikoSRCalculateService implements MessageService<Matcher> {
    private static final char CHAR_X = 'x';
    private static final char CHAR_O = 'o';
    private static final char CHAR_NONE = '-';
    enum status{
        XX,OO,XO,OX,X,O,NONE;
        static status get(char a, char b){
            if (a == CHAR_X){
                if (b == CHAR_X) return XX;
                else if (b == CHAR_O) return XO;
                else if (b == CHAR_NONE) return X;
            } else if (a == CHAR_O) {
                if (b == CHAR_X) return OX;
                else if (b == CHAR_O) return OO;
                else if (b == CHAR_NONE) return O;
            } else {
                if (b == CHAR_X) return X;
                else if (b == CHAR_O) return O;
                else if (b == CHAR_NONE) return NONE;
            }
            return NONE;
        }
    }

    @Override
    public boolean isHandle(MessageEvent event, String messageText, DataValue<Matcher> data) {
        var m = Instructions.TEST_TAIKO_SR_CALCULATE.matcher(messageText);
        if (m.find()) {
            data.setValue(m);
            return true;
        } else return false;
    }

    @Override
    public void HandleMessage(MessageEvent event, Matcher matcher) throws Throwable {
        var xtr = matcher.group("data").trim().replaceAll("\\s+",String.valueOf(CHAR_NONE));
        event.getSubject().sendMessage("结果:"+get(xtr));
    }

    private float get(String xtr){
        int nexIndex = 4;
        LinkedList<status> statusList = new LinkedList<>();
        var datas = new char[xtr.length()];xtr.getChars(0, xtr.length(), datas, 0);
        //处理前4个
        if (datas.length <= 3) {
            //?
        } else if (datas[1] == CHAR_NONE) {
            statusList.add(datas[0] == CHAR_X ? status.XX : status.OO);
            statusList.add(status.NONE);
            statusList.add(status.get(datas[2], datas[3]));
            nexIndex = 4;
        } else if (datas[2] == CHAR_NONE){
            statusList.add(status.get(datas[0], datas[1]));
            statusList.add(status.NONE);
            nexIndex = 3;
        } else if (datas[3] == CHAR_NONE) {
            statusList.add(status.get(datas[0], datas[1]));
            statusList.add(status.get(datas[2], datas[3]));
            statusList.add(status.get(datas[1], datas[2]));
            statusList.add(status.NONE);
            nexIndex = 4;
        } else {
            statusList.add(status.get(datas[0], datas[1]));
            nexIndex = 2;
        }
        //

        for (int i = nexIndex; i < datas.length; i += nexIndex) {
            if ((i+3) >= datas.length) {
                nexIndex = i;
                break;
            }
            nexIndex = checkNext(statusList, datas[i], datas[i+1] , datas[i+2], datas[i+3]);
        }
        int t = datas.length - nexIndex;
        if (t == 1){
            statusList.add(status.get(datas[nexIndex], CHAR_NONE));
        } else if (t == 2){
            statusList.add(status.get(datas[nexIndex], datas[nexIndex+1]));
        } else if (t == 3){
            statusList.add(status.get(datas[nexIndex], datas[nexIndex+1]));
            statusList.add(status.get(datas[nexIndex+2], CHAR_NONE));
        }


        //解析
        status thisStatus = statusList.poll();
        status nextStatus;

        int a = 0;
        int b = 0;
        int c = 0;
        while ((nextStatus = statusList.poll()) != null){
            if (nextStatus == status.NONE) {
                a++;
                b++;
                break;
            }
            switch (thisStatus){
                case OO:{
                    switch (nextStatus){
                        case OX -> {b++;c++;thisStatus = nextStatus;}
                        case XO -> {a++;c++;thisStatus = nextStatus;}
                        case XX -> {b++;a++;thisStatus = nextStatus;}
                        case X -> a++;
                    }
                } break;
                case XX:{
                    switch (nextStatus){
                        case OX -> {a++;c++;thisStatus = nextStatus;}
                        case XO -> {b++;c++;thisStatus = nextStatus;}
                        case OO -> {b++;a++;thisStatus = nextStatus;}
                        case O -> a++;
                    }
                } break;
                case XO:{
                    switch (nextStatus){
                        case OX -> {a++;b++;thisStatus = nextStatus;}
                        case XX -> {b++;c++;thisStatus = nextStatus;}
                        case OO -> {a++;c++;thisStatus = nextStatus;}
                        case X -> c++;
                        case O -> {
                            a++;
                            c++;
                        }
                    }
                } break;
                case OX:{
                    switch (nextStatus){
                        case XO -> {a++;b++;thisStatus = nextStatus;}
                        case XX -> {a++;c++;thisStatus = nextStatus;}
                        case OO -> {b++;c++;thisStatus = nextStatus;}
                        case O -> c++;
                        case X -> {
                            a++;
                            c++;
                        }
                    }
                }
            }
        }
        return (0.25f*a + b) * (0.5f*c + 1);
    }

    private int checkNext(List<status> statusList, char a, char b, char c, char d){
        if (c == CHAR_NONE){
            statusList.add(status.get(a, b));
            statusList.add(status.NONE);
            return 3;
        } else if (d == CHAR_NONE) {
            statusList.add(status.get(a, b));
            statusList.add(status.get(c, d));
            statusList.add(status.get(b, c));
            statusList.add(status.NONE);
            return 4;
        } else {
            statusList.add(status.get(a, b));
            return 2;
        }
    }
}

File: src/main/java/com/now/nowbot/service/messageServiceImpl/TestTaikoSRCalculateService.kt
package com.now.nowbot.service.messageServiceImpl

import com.now.nowbot.qq.event.MessageEvent
import com.now.nowbot.service.MessageService
import com.now.nowbot.service.MessageService.DataValue
import com.now.nowbot.util.Instruction
import org.springframework.stereotype.Service
import java.util.regex.Matcher

@Service("TEST_TAIKO_SR_CALCULATE") 
class TestTaikoSRCalculateService : MessageService<Matcher> {
    
    override fun isHandle(event: MessageEvent, messageText: String, data: DataValue<Matcher>): Boolean {
        val m = Instruction.TEST_TAIKO_SR_CALCULATE.matcher(messageText)
        if (m.find()) {
            data.value = m
            return true
        } else return false
    }
    
    @Throws(Throwable::class) override fun HandleMessage(event: MessageEvent, matcher: Matcher) {
        val str = matcher.group("data").trim {it <= ' '} .replace("\\s+".toRegex(), "-")
        event.reply("结果：" + getResult(str))
    }
    
    private fun getResult(str: String): String {
        var mainColorChanged = 0
        var mainFingerChanged = 0
        var subColorChanged = 0
        var subFingerChanged = 0
        var complexChanged = 0
        var complexSameColor = 0

        var isComplex = false
        var complexCount = 0

        for (i in str.indices) {
            val char = str[i]

            if (isComplex.not() && (char == '(' || char == '（')) {
                isComplex = true
                complexChanged ++
                continue
            } else if (isComplex && (char == ')' || char == '）')) {
                isComplex = false
                complexCount = 0
                continue
            }

            val index = if (isComplex) complexCount else i

            val frontChar = if (index >= 1) str[index - 1] else '-'
            val front2Char = if (index >= 2) str[index - 2] else '-'

            if (frontChar.toString().matches("[xo]".toRegex()) && char != '-' && frontChar != char) {
                if (i % 2 == 0) mainColorChanged ++ else subColorChanged ++
            }

            if (front2Char.toString().matches("[xo]".toRegex()) && char != '-' && front2Char != char) {
                if (i % 2 == 0) mainFingerChanged ++ else subFingerChanged ++
            }

            if (frontChar.toString().matches("[()（）]".toRegex()) && front2Char.toString().matches("[xo]".toRegex()) && front2Char == char) {
                complexSameColor ++
            }

            if (isComplex) complexCount ++
        }

        val isStartDifferent = if (str.length > 1) {
            str.first() != str.last()
        } else false

        val difficulty = mainColorChanged + mainFingerChanged + 3 * ((subColorChanged + 1) * (subFingerChanged + 1) - 1) + (if (isStartDifferent) 1 else 0)

        return """
            对 $str 的难度分析：
            主手换色：$mainColorChanged
            主手换起指：$mainFingerChanged
            副手换色：$subColorChanged
            副手换起指：$subFingerChanged
            
            预期难度：$difficulty
            
            复合：$complexChanged
            复合前后同色：$complexSameColor
        """.trimIndent()
    }
}
